kernel/src/hil/asymmetric_encryption.rs

```diff
@@ -0,0 +1,133 @@
+//! Interface for Public/Private key encryption
+
+use crate::hil::entropy;
+use crate::ErrorCode;
+
+/// Upcall from the `PubPrivKey` trait.
+pub trait PubPrivKeyClient<'a> {
+    /// The `generate()` command has been completed.
+    fn generation_complete(&'a self, result: Result<(), ErrorCode>);
+}
+
+/// An internal representation of a asymetric Public and/or Private key.
+///
+/// This trait is useful for managing keys internally in Tock.
+pub trait PubPrivKey<'a> {
+    /// Set the client. This client will be called when the `generate()`
+    /// function is complete. If using an existing key this doesn't need to be
+    /// used.
+    fn set_client(&'a self, client: &'a dyn PubPrivKeyClient<'a>);
+
+    /// This generates a new private/public key pair of `length` bits.
+    /// This will call the `generation_complete()` on completion. They keys
+    /// cannot be used and will return `None` until the upcall has been called.
+    ///
+    /// The keys generated by `generate()` will depend on the implementation.
+    fn generate(&'a self, trng: &'a dyn entropy::Entropy32, length: usize)
+        -> Result<(), ErrorCode>;
+
+    /// Import an existing public key. This will override any existing keys.
+    fn import_public_key(&mut self, public_key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Import an existing private key. This will override any existing keys.
+    fn import_private_key(&mut self, private_key: &[u8]) -> Result<(), ErrorCode>;
+
+    /// Store the public key in buffer if it exists.
+    fn pub_key(&'a self, buffer: &mut [u8]) -> Option<()>;
+
+    /// Store the private key in buffer if it exists.
+    fn priv_key(&'a self, buffer: &mut [u8]) -> Option<()>;
+}
+
+pub enum Operation {
+    Encrypt,
+    Decrypt,
+    Sign,
+    Verify,
```

Comment:
It seems weird that signatures and signature verification are in a file about `asymmetric_encryption`. Perhaps this should be called `public_key_crypto.rs`? That would then mean we could expect crypto that's just signatures in here as well. Otherwise, it might make sense separate these out into a separate `signatures.rs`.